### ISSUE TITLE
Redis cache: fwrite(): send of 8192 bytes failed with errno=104 Connection reset by peer

### ISSUE THREAD

**Issue by ilnytskyi** [magento/magento2#27151]
Recently we noticed a new error appearing in different places like:
- Product save in admin
- cache:clean
- cleaning cache by the indexer
```
Notice: fwrite(): send of 8192 bytes failed with errno=104 Connection reset by peer in /var/www/html/vendor/colinmollenhour/credis/Client.php on line 1229

```
### Preconditions (*)

1. use any magento multistore with a lot of products that generate many different keys
2. just many keys in Redis

that was hard to debug however we found something
### Steps to reproduce (*)

1. Write a lot of keys into Redis in my case it was 800K - 1M keys()
```
/** @var \Magento\Framework\App\CacheInterface $cache */
$cache = $obj->get(\Magento\Framework\App\CacheInterface::class);

echo 'REDIS TRANSFER';

for ($i = 0; $i < 1000000; $i++) {
    $cache->save('some_data', 'MY_KEY_' . $i);
}
```
2. Clean it using this method. (probably used by indexer or some extensions) hard to reproduce in live store and locate the place.
```
$cache->getFrontend()->clean();
```
3. See error
![Selection_445](https://user-images.githubusercontent.com/3192770/75866711-c0c1df80-5e05-11ea-967b-87d9947e11e9.png)
4. Cleaning cache this way generates an error `-ERR Protocol error: invalid multibulk length`

Probably related to the max length that  Redis can receive
https://github.com/antirez/redis/blob/6c60526db91e23fb2d666fc52facc9a11780a2a3/src/networking.c#L1023-L1027

### Expected result (*)

1. When total keys length is greater than `1024 * 1024 = 1048576` their clean should be processed in batches or someway improved
2. Cache cleaned and this problem does not affect Magento app functions

### Actual result (*)

1. When total keys length is greater than `1024 * 1024 = 1048576` their clean fails and generate an error
2. The cache is not cleaned and this problem affects Magento app

PS:
It looks more like Redis lib issue however Magento may not handle this case correctly and overuse clean by id
Reported here as well and it can be used as a workaround with optimal batch size up to 10K: https://github.com/colinmollenhour/Cm_Cache_Backend_Redis/issues/153

Additionally, improve cache of the swatches block as here (will be in 2.4) or better 
https://github.com/magento/magento2/commit/ccc4301356130c4bac27ab99e2fe4f1fdcba8ec3#diff-4722825b1709a7da93ee9115244a0a09

**Comment by onlinebizsoft** [comment]
This is a real issue and it is still valid, please reopen @sdzhepa @sidolov @ilnytskyi

Either update lib or provide a fix in Magento

**Comment by hostep** [comment]
@onlinebizsoft: I believe this was done in: https://github.com/magento/magento2/pull/33364, I'm just not sure when this change will get delivered, in which Magento version ...

**Comment by onlinebizsoft** [comment]
@hostep you are right. it is bad when this was pointed out since March 2020 and until now, it is not included in 2.4.3
